README.md

# WebMixer

A library for scraping urls

## The Basic Scraper

All `webmixer.scrapers.pages` and `webmixer.scrapers.tags` classes inherit from `webmixer.base.BasicScraper`, which means they all have the following attributes and functions:

### Attributes
* directory (str): Directory to write files to
* color (str): Color for error messages (default: 'rgb(153, 97, 137)')
* locale (str): Language to use when writing error messages (default: 'en')
	* Note: must be listed in `webmixer.messages.MESSAGES`
* default_ext (str): Extension to default to for extracted files

### Functions
#### create_tag(tag)
Args:
* tag (str): tag name to create (e.g. 'p')

Returns a BeautifulSoup tag
Example:
```
image_tag = create_tag('img')
```
___

#### get_filename(link, default_ext=None)

Args:
* link (str): URL that has been scraped
* default_ext (optional str): if the link doesn't have an extension, use this extension'

Returns a filename (str) to use for extracted files
Example:
```
video_filename = get_filename('<url>', default_ext='.mp4')
```
___

#### mark_tag_to_skip(tag)
Mark tag to skip during further scraping operations
Args:
* tag (str): tag to mark

Example:
```
Process img tag here...

mark_tag_to_skip(img)
```
___

#### write_url(link, url=None, default_ext=None, filename=None, directory=None)
Downloads a url and writes it to a zip
Args:
* filepath (str): path to local file
* directory (str): directory to write to zip
* url (optional str): URL used for handling relative URLs
* default_ext (optional str): if the link doesn't have an extension, use this extension
* filename (optional str): name for file to write to zip
* directory (optional str): directory to write file to zip

Returns filepath within zip
Example:
```
write_url('<link>', url='https://domain.com/', default_ext='.mp4', filename='video', directory='media') # 'media/video.mp4'
```
___

#### write_contents(filename, contents, directory=None)
Writes contents to the zip with a given filename
Args:
* filename (str): filename for contents
* contents (bytes): contents to write to zip
* directory (str): directory to write to zip

Returns filepath within zip
Example:
```
write_contents('myfile.pdf', <pdf contents>, directory='docs')  # docs/myfile.pdf
```
___

#### write_file(filepath, directory=None)
Writes a local file to the zip
Args:
* filepath (str): path to local file
* directory (str): directory to write to zip

Returns filepath within zip
Example:
```
write_file('path/to/myfile.mp3', directory='music')  # music/myfile.mp3
```
___


#### create_broken_link_message(link)
Generates a tag with broken link message
Args:
* link (str): link to copy/paste

Returns a div tag with a link to copy/paste into browser
Example:
```
iframe.replaceWith(create_broken_link_message('<url>'))
# iframe -> <div>copy link...</div>
```
___

#### create_copy_link_message(link, partially_scrapable=False)
Generates a tag with 'copy link into browser' message
Args:
* link (str): link to copy/paste
* partially_scrapable (bool): link was mostly scraped, but doesn't include everything from original site

Returns a div tag with a link to copy/paste into browser
Example:
```
iframe.replaceWith(create_copy_link_message('<url>'))
# iframe -> <div>copy link...</div>
```
___

### Exceptions
`webmixer.exceptions` can be useful for handling errors from a variety of sources. If you are scraping a more specialized source, there may be some exceptions that are exclusive to that source. You can then raise the following exceptions to correctly manage that source:

#### BrokenSourceException
Used when the link is completely broken (e.g. site no longer exists)

#### UnscrapableSourceException
Used when the link is working, but cannot be supported on Kolibri (e.g. Flash content)

For instance, the `webmixer.scrapers.pages.gdrive.GoogleDriveScraper` may throw a `FileNotDownloadableError` error. In order to handle this correctly, it will raise an `UnscrapableSourceException`
```
try:
	...
except FileNotDownloadableError as e:
	raise UnscrapableSourceException(e)
```


## Page Scrapers
There are several page scrapers that are available for use in scraping html pages. These will download urls to their respective file types

### Built-in Scrapers
Here is a list of the basic scraper classes, which are also listed under `webmixer.scrapers.pages.base.COMMON_SCRAPERS`:
*  WebVideoScraper
* PDFScraper
* EPubScraper
* ImageScraper
* FlashScraper
* VideoScraper
* AudioScraper


### Using Page Scrapers
When you create a scraper object, you may specify the following:
* url (str): URL that tag can be found at (used to handle relative URLs) __required__
* zipper (optional `ricecooker.utils.html_writer`): Zip to write to
* triaged (optional [str]): List of already parsed URLs

To scrape the page, you may use any of the following writing options:

__to_zip__: Writes a file to self.zipper, which is useful when scraping embedded sources from an html page
Args:
* filename (optional str): name of file to write to
Returns path to file from within zip

Here are the default extensions for each `webmixer.scrapers.pages.base.Scraper`:
| Scraper | Extension |
|--|--|
| HTMLPageScraper  |.html |
|PDFScraper|.pdf|
| EPubScraper | .epub |
| AudioScraper | .mp3 |
| VideoScraper | .mp4 |
| WebVideoScraper | .mp4 |
|ImageScraper| .png |
|FlashScraper| _error_ |

For example:
```
from webmixer.scrapers.base import ImageScraper
image= <BeautifulSoup tag>
image['src'] = ImageScraper('<url>').to_zip()  # Sets 'src' to zipped image filepath
```
___

__to_tag__: Writes file to zip and generates a tag based on what kind of scraper it is. This is useful when you are replacing iframes with native html elements
Args:
* filename (optional str): name of file to write to
Returns tag

Here are the return tag types for each `webmixer.scrapers.pages.base.Scraper`:
| Scraper | Tag |
|--|--|
| HTMLPageScraper  | None|
|PDFScraper|\<embed\>|
| EPubScraper | None |
| AudioScraper | \<audio\> |
| VideoScraper | \<video\> |
| WebVideoScraper | \<video\> |
|ImageScraper|\<img\>|
|FlashScraper|_error_|

For example:
```
from webmixer.scrapers.base import PDFScraper
iframe= <BeautifulSoup tag>
iframe.replaceWith(PDFScraper('<url>').to_tag())  # Replaces iframe with <embed> tag
```

___

__to_file__: Writes to a file. This is useful for downloading URLs as files to your local machine.
Args:
* filename (optional str): name of file to write to
* directory (optional str): directory to write to
* overwrite (bool): overwrite file if it exists
Returns a filepath to the downloaded file

`to_file` uses the `download_file` method to write the file to a `write_to_path`

Here are the return file types for each `webmixer.scrapers.pages.base.Scraper`:
| Scraper | Extension |
|--|--|
| HTMLPageScraper  | .zip - generated by `ricecooker.utils.html_writer` |
|PDFScraper|.pdf |
| EPubScraper | .epub |
| AudioScraper | .mp3 |
| VideoScraper | .mp4 |
| WebVideoScraper | .mp4 |
|ImageScraper|_error - content kind not supported_  |
|FlashScraper|_error_ |

For example:
```
from webmixer.scrapers.base import HTMLPageScraper
new_html_zip_path = HTMLPageScraper('<url>').to_file() # Returns newly scraped html .zip file
```

### Custom Scrapers
Given how diverse the internet is, you may need to implement your own scraper to handle individual sources. You __must__ implement a `test` classmethod in order to use your scraper.

__If you would like to share a custom scraper, please feel free to open a pull request with a new file under `webmixer.scrapers.pages`__

#### Attributes
All scrapers have the following attributes:
* dl_directory (str): Directory to write `to_file` downloaded file to (default: 'downloads')
* directory (str): Directory to write files to
* color (str): Color for error messages (default: 'rgb(153, 97, 137)')
* locale (str): Language to use when writing error messages (default: 'en')
	* Note: must be listed in `webmixer.messages.MESSAGES`
* default_ext (str): Extension to default to for extracted files
* kind (`le_utils.constants.content_kind`): Content kind to write to

`webmixer.scrapers.pages.base.HTMLPageScraper` has these additional attributes:
* partially_scrapable (bool): Not all content can be viewed from within Kolibri (default: False)
* scrape_subpages (bool): Determines whether to scrape any subpages within this page (default: True)
* main_area_selector (optional tuple): Main element selector to replace everything in body tag
* omit_list (optional list): list of selectors to remove from page contents (e.g. [('a', {'class': 'link'})])
* loadjs (bool): Determines whether to load js when loading the page (default: True)
* scrapers ([`webmixer.scrapers.pages.Scraper`]): List of additional scrapers to use on this page
* extra_tags ([`webmixer.scrapers.tags.Tag`]): List of additional tags to scrape

For example, the following code will remove links, scrape Wikipedia pages, and sets all images to 'myimg.png':
```
from webmixer.scrpaers.tags import ImageTag
from webmixer.scrapers.pages.base import HTMLPageScraper
from webmixer.scrapers.pages.wikipedia import WikipediaScraper

class MyCustomTag(ImageTag):
	def process(self):
		self.tag['src'] = self.write_file('myimg.png')

class MyCustomScraper(HTMLPageScraper):
	omit_list = [('a',)]  		   # Remove links
	extra_tags = [MyCustomTag]     # Use MyCustomTag to set images to 'myimg.png'
	scrapers = [WikipediaScraper]  # Scrape any Wikipedia pages

	@classmethod                   # Required test classmethod
	def test(self, url):
		return 'my-domain.com' in url
```

#### Functions
__@classmethod test(url)__: Required method to determine if this is the correct scraper for this URL
Args:
  * url (str): url to test
Returns True if scraper is meant to scrape URL
Example:
```
@classmethod
def test(self, url):
	return 'somedomain' in url
```

___

__preprocess(contents)__: Process contents before main scraping method
Args:
	contents (BeautifulSoup): contents to preprocess
Example:
```
# Delete the first image on the page before scraping all the images
def preprocess(self, contents):
	contents.find('img').decompose()
```
___

__postprocess(contents)__: Process contents after main scraping method
Args:
	contents (BeautifulSoup): contents to postprocess
Example:
```
# Append a link at the end of the <body> tag
def postprocess(self, contents):
	link = self.create_tag('a')
	link.string = 'New Link'
	contents.body.append(link)
```


## Tags
There are several tags that are available for use in scraping html pages. These will handle downloading any referenced files.

### Using Tags
To create a tag, you may specify the following:
* tag (BeautifulSoup.tag): tag to parse __required__
* url (str): url that tag can be found at (used to handle relative URLs) __required__
* attribute (optional str): attribute to find link at (e.g. 'src' or 'data-src')
* scrape_subpages (optional bool): parse linked pages referenced by this tag (default: True)
* extra_scrapers (optional [`webmixer.scrapers.base.BasicScrapers`]): list of scrapers to try to scrape linked pages
* color (optional str): color for injected error messages (default: 'rgb(153, 97, 137)')

To scrape the tag, use the `scrape` method. This will process the tag so that it can be usable from within an html zip. Here is a simple scraping example:
```
from webmixer.scrapers.tags import ImageTag
image_tag = <BeautifulSoup.img tag>
image_scraper = ImageTag(image_tag, '<url>')
image_scraper.scrape()  # image_tag['src'] will point to downloaded image file in zip
```



### Built-in Tags
Here is a list of the available tags, which are also listed under `webmixer.scrapers.tags.COMMON_TAGS`

* ImageTag (img)
* AudioTag (audio)
* VideoTag (video)
* EmbedTag (embed)
* LinkTag (a) _Scrapes linked pages referenced by 'href' attribute_
* IframeTag (iframe) _Scrapes embedded pages referenced byon 'src' attribute_
* StyleTag (style) _Scrapes sheets referenced by 'href' attribute_
* ScriptTag (script) _Scrapes scripts referenced by 'src' attribute_


### Custom Tags
Depending on the source you are trying to scrape, you may need more specific methods for scraping a page. To create a custom tag, you will need to subclass `webmixer.scrapers.tags.BasicScraperTag`

#### Attributes
All tags have the following attributes:
* selector (tuple): BeautifulSoup selector to find tag (e.g. ('a', {'class': 'link'}))
* default_ext (str): Extension to use if link doesn't have an extension
* directory (str): Directory to write tag files to
* attributes (dict): Any attributes to assign to a tag
* default_attribute (str): Attribute that references files (default: 'src')
* scrape_subpages (bool): Determines whether to scrape any linked pages (default: True)
* extra_scrapers ([`webmixer.scrapers.base.BasicScrapers`]): List of additional scrapers to use for scraping linked pages
* color (str): Color for error messages (default: 'rgb(153, 97, 137)')
* locale (str): Language to use when writing error messages (default: 'en')
	* Note: must be listed in `webmixer.messages.MESSAGES`

Example:
```
from webmixer.scrapers.tags import BasicScraperTag

class MyVideoTag(BasicScraperTag):
	selector = ('video', {'class': 'video-class'})  # Select video.video-class
	directory = 'media'								# Files will be written to media folder
	attributes = {									# Videos will have width 100%
		'width': '100%'
	}
```


#### Built-in functions
For more custom scraping logic, you may also override the following methods:

__process()__: Makes the tag usable from within an html zip by downloading any referenced files
Example:
```
class MyVideoTag(BasicScraperTag):
	def process(self):
		# Scrape all of the <source> tags
	    for source in self.tag.find_all('source'):
	        BasicScraperTag(source, self.zipper, self.url).scrape()
```
___

__handle_error()__: Determines how to handle cases where the link is broken
Example:
```
class MyVideoTag(BasicScraperTag):
	def handle_error(self):
		self.tag.decompose()  # Just remove the element if it doesn't work
```
___

__handle_unscrapable()__: Determines how to handle cases where the link is not scrapable
Example:
```
class MyVideoTag(BasicScraperTag):
	def handle_unscrapable(self):
		self.tag.replaceWith(self.create_copy_link_message(self.link))
```


## Helper Functions
### webmixer.utils.guess_scraper
If you would like to determine which scraper to use based on a URL, you can use the `webmixer.utils.guess_scraper` method. This will accept the following arguments:
* url (str): URL to scrape
* scrapers ([`webmixer.scrapers.base.BasicScrapers`]): list of other scrapers to test URL against
* allow_defualt (optional bool): use generic default scraper in case nothing matches (default: False)

_You can also pass in additional arguments to scrapers with `kwargs`_

So a simple usage of `guess_scraper` might be:
```
from webmixer.utils import guess_scraper
scraper = guess_scraper('<url>', scrapers=[MyCustomScraper])
```


docs/history.rst
=======
History
=======

0.0.0 (2019-07-30)
------------------
* First release on PyPI.



webmixer/scrapers/pages/base.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-
import os
from bs4 import BeautifulSoup
from ricecooker.utils import downloader, html_writer
from ricecooker.config import LOGGER              # Use LOGGER to print messages
import youtube_dl
import shutil
import tempfile

from le_utils.constants import content_kinds

from webmixer.exceptions import EXCEPTIONS, UnscrapableSourceException
from webmixer.scrapers.base import BasicScraper
from webmixer.scrapers.tags import COMMON_TAGS


class BasicPageScraper(BasicScraper):
    dl_directory = 'downloads'

    @classmethod
    def test(self, url):
        """
            Used to determine if this is the correct scraper to use for a given url
            Args:
                url (str): url to test
            Returns True if scraper is meant to scrape URL
        """
        raise NotImplementedError('Must implement a test method for {}'.format(str(self.__class__)))


    def preprocess(self, contents):
        """
            Place for any operations to occur before main scraping method
            Args:
                contents (BeautifulSoup): contents of page to pre-process
        """
        # Implement in subclasses
        pass

    def process(self):
        return downloader.read(self.url)

    def postprocess(self, contents):
        """
            Place for any operations to occur after main scraping method
            Args:
                contents (BeautifulSoup): contents of page to post-process
        """
        # Implement in subclasses
        pass


    ##### Output methods #####
    def download_file(self, write_to_path):
        """
            Downloads file to write_to_path
            Args:
                write_to_path (str): path to write file to
        """
        with open(write_to_path, 'wb') as fobj:
            fobj.write(self.process())

    def to_file(self, filename=None, directory=None, overwrite=False):
        """
            Download URL as a file
            Args:
                filename (optional str): name of file to write to
                directory (optional str): directory to write to
                overwrite (bool): overwrite file if it exists
        """
        directory = directory or self.directory
        if not os.path.exists(directory):
            os.makedirs(directory)

        write_to_path = os.path.join(directory, filename or self.get_filename(self.url))

        if overwrite or not os.path.exists(write_to_path):
            self.download_file(write_to_path)

        return write_to_path

    def to_zip(self, filename=None):
        """
            Download URL to self.zipper
            Args:
                filename (optional str): name of file to write to
        """
        return self.write_url(self.url, filename=filename)

    def to_tag(self, filename=None):
        """
            Download URL and return a tag based on what kind of scraper this is
            Args:
                filename (optional str): name of file to write to
        """
        raise NotImplementedError('Must implement to_tag function on {}'.format(str(self.__class__)))


######### HTML SCRAPERS ##########

class HTMLPageScraper(BasicPageScraper):
    partially_scrapable = False     # Not all content can be viewed from within Kolibri (e.g. Wikipedia's linked pages)
    scrape_subpages = True          # Determines whether to scrape any subpages within this page
    default_ext = '.html'           # Default extension when writing files to zip
    main_area_selector = None       # Place where main content is (replaces everything in <body> with this)
    omit_list = None                # Specifies which elements to remove from the DOM
    loadjs = False                  # Determines whether to load js when loading the page
    scrapers = None                 # List of additional scrapers to use on this page (e.g. GoogleDriveScraper)
    extra_tags = None               # List of additional tags to look for (e.g. ImageTag)
    color = 'rgb(153, 97, 137)'     # Color to use for messages (consider contrast when setting this)
    kind = content_kinds.HTML5      # Content kind to write to


    def __init__(self, *args, **kwargs):
        super(HTMLPageScraper, self).__init__(*args, **kwargs)
        self.omit_list = self.omit_list or []
        self.omit_list += [
            ('link', {'type': 'image/x-icon'}),
            ('link', {'rel': 'apple-touch-icon'}),
            ('span', {'class': 'external-iframe-src'}),
            ('link', {'rel': 'icon'}),
            ('div', {'class': 'mw-indicators'})
        ]
        self.extra_tags = self.extra_tags or []
        self.scrapers = (self.scrapers or []) + [self.__class__]

    def process(self):
        # Using html.parser as it is better at handling special characters
        contents = BeautifulSoup(downloader.read(self.url, loadjs=self.loadjs), 'html.parser')

        self.preprocess(contents)

        # If a main area is specified, replace body contents with main area
        if self.main_area_selector:
            body = self.create_tag('body')
            body.append(contents.find(*self.main_area_selector))
            contents.body.replaceWith(body)

        # Remove any items to omit
        for item in self.omit_list:
            for element in contents.find_all(*item):
                element.decompose()

        # Scrape tags
        for tag_class in (self.extra_tags + COMMON_TAGS):
            for tag in contents.find_all(*tag_class.selector):
                scraper = tag_class(tag, self.url,
                    zipper=self.zipper,
                    scrape_subpages=self.scrape_subpages,
                    triaged=self.triaged,
                    locale=self.locale,
                    extra_scrapers=self.scrapers,
                    color=self.color
                )
                scraper.scrape()

        self.postprocess(contents)

        return contents.prettify(formatter="minimal").encode('utf-8-sig', 'ignore')

    ##### Output methods #####
    def download_file(self, write_to_path):
        # Generate a .zip file
        with html_writer.HTMLWriter(write_to_path) as zipper:
            try:
                self.zipper = zipper
                self.to_zip(filename='index.html')
            except Exception as e:
                # Any errors here will just say index.html file does not exist, so
                # print out error for more descriptive debugging
                LOGGER.error(str(e))

    def to_file(self, filename=None, **kwargs):
        # Make sure html is being written to a zip file here
        filename = filename or self.get_filename(self.url)
        filename = filename.replace(self.default_ext, '.zip')
        return super(HTMLPageScraper, self).to_file(filename=filename, **kwargs)

    def to_zip(self, filename=None):
        return self.write_contents(filename or self.get_filename(self.url), self.process())


class DefaultScraper(HTMLPageScraper):
    """ Basic HTML page scraper in case no other scrapers match the URL """
    scrape_subpages = False
    loadjs = True

    def __init__(self, *args, **kwargs):
        super(DefaultScraper, self).__init__(*args, **kwargs)
        self.scrapers = [] # Don't scrape any subpages

    @classmethod
    def test(self, url):
        ext = os.path.splitext(url.split('?')[0].split('#')[0])[1].lower()
        return not ext or ext.startswith('.htm')


########## OTHER CONTENT KIND SCRAPERS ##########

class ImageScraper(BasicPageScraper):
    directory = 'img'
    default_ext = '.png'
    kind = content_kinds.SLIDESHOW  # No image type in Studio, so use this

    @classmethod
    def test(self, url):
        return url.lower().endswith('.png') or url.lower().endswith('.jpg')

    def to_file(self, filename=None):
        raise NotImplementedError('Unable to write SLIDESHOW kind to a file')

    def to_tag(self, filename=None):
        try:
            img = self.create_tag('img')
            img['src'] = self.to_zip(filename=filename)
            return img
        except EXCEPTIONS as e:
            LOGGER.error(str(e))
            return self.create_broken_link_message(self.url)


class PDFScraper(BasicPageScraper):
    directory = 'docs'
    default_ext = '.pdf'
    kind = content_kinds.DOCUMENT

    @classmethod
    def test(self, url):
        return url.split('?')[0].lower().endswith('.pdf')

    def to_tag(self, filename=None):
        try:
            embed = self.create_tag('embed')
            embed['src'] = self.to_zip(filename=filename)
            embed['width'] = '100%'
            embed['style'] = 'height: 500px;max-height: 100vh;'
            return embed
        except EXCEPTIONS as e:
            LOGGER.error(str(e))
            return self.create_broken_link_message(self.url)

class EPubScraper(BasicPageScraper):
    directory = 'docs'
    default_ext = '.epub'
    kind = content_kinds.DOCUMENT

    @classmethod
    def test(self, url):
        return url.split('?')[0].lower().endswith('.epub')


class AudioScraper(BasicPageScraper):
    default_ext = '.mp3'
    kind = content_kinds.AUDIO
    directory = 'audio'

    @classmethod
    def test(self, url):
        return url.split('?')[0].lower().endswith('.mp4')

    def to_tag(self, filename=None):
        try:
            audio = self.create_tag('audio')
            audio['controls'] = 'controls'
            audio['style'] = 'width: 100%;'
            source = self.create_tag('source')
            source['src'] = self.to_zip(filename=filename)
            audio.append(source)
            return audio
        except EXCEPTIONS as e:
            LOGGER.error(str(e))
            return self.create_broken_link_message(self.url)


class VideoScraper(BasicPageScraper):
    default_ext = '.mp4'
    kind = content_kinds.VIDEO
    dl_directory = 'videos'
    directory = 'videos'

    @classmethod
    def test(self, url):
        return url.split('?')[0].lower().endswith('.mp4')

    def to_tag(self, filename=None):
        try:
            video = self.create_tag('video')
            video['controls'] = 'controls'
            video['style'] = 'width: 100%;'
            video['preload'] = 'auto'
            source = self.create_tag('source')
            source['src'] = self.to_zip(filename=filename)
            video.append(source)
            return video
        except EXCEPTIONS as e:
            LOGGER.error(str(e))
            return self.create_broken_link_message(self.url)


class WebVideoScraper(VideoScraper):
    """ Videos that can be scraped by youtube_dl """

    @classmethod
    def test(self, url):
        return 'youtube' in url or 'vimeo' in url

    def process(self):
        write_to_path = self.to_file()
        with open(write_to_path) as fobj:
            return fobj.read()

    def download_file(self, write_to_path):
        try:
            dl_settings = {
                'outtmpl': write_to_path,
                'quiet': True,
                'overwrite': True,
                'format': self.default_ext.split('.')[-1],
            }
            with youtube_dl.YoutubeDL(dl_settings) as ydl:
                ydl.download([self.url])
        except (youtube_dl.utils.DownloadError, youtube_dl.utils.ExtractorError) as e:
            raise UnscrapableSourceException(str(e))  # Some errors are region-specific, so allow link

    def to_zip(self, filename=None):
        try:
            tempdir = tempfile.mkdtemp()
            video_path = os.path.join(tempdir, filename or self.get_filename(self.url))
            self.download_file(video_path)
            return self.write_file(video_path)
        except FileNotFoundError as e:
            # Some video links don't work, so youtube dl only partially downloads files but doesn't error out
            # leading to the .mp4 not being found (just a .part file)
            raise UnscrapableSourceException(str(e))
        finally:
            shutil.rmtree(tempdir)

    def to_tag(self, filename=None):
        video = self.create_tag('video')
        video['controls'] = 'controls'
        video['style'] = 'width: 100%;'
        video['preload'] = 'auto'
        source = self.create_tag('source')
        source['src'] = self.to_zip(filename=filename)
        video.append(source)
        return video


class FlashScraper(BasicPageScraper):
    default_ext = '.swf'
    standalone = True
    kind = 'flash'

    @classmethod
    def test(self, url):
        return url.split('?')[0].lower().endswith('.swf')

    def process(self, **kwargs):
        # Automatically throw error
        downloader.read(self.url) # Raises broken link error if fails
        raise UnscrapableSourceException('Cannot scrape Flash content')

    def to_tag(self, **kwargs):
        return self.process()

    def to_zip(self, **kwargs):
        return self.process()


COMMON_SCRAPERS = [
    WebVideoScraper,
    PDFScraper,
    EPubScraper,
    ImageScraper,
    FlashScraper,
    VideoScraper,
    AudioScraper,
]
